### PR TITLE
Add short cut for recomputed tv

### LINKED ISSUE #2094: Codegen failure in HuggingFace AllenaiLongformerBase with Autocast
### 🐛 Describe the bug

Benchmark command line:

```
python -u benchmarks/huggingface.py --training -d cuda --fast --backend nvprims_nvfuser --skip-accuracy-check --performance --only AllenaiLongformerBase --amp
```

Repro:

```
import torch
from torch._C._nvfuser import FusionDefinition, Fusion, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float)     
    T1 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float)
    T2 = fd.define_tensor(symbolic_sizes=[-1, -1], contiguous=[True, True], dtype=DataType.Half)                                                                                                                                                     
    T3 = fd.ops.broadcast_in_dim(T0, output_shape=[1, 1024, 768], broadcast_dims=[2])
    T4 = fd.ops.broadcast_in_dim(T1, output_shape=[1, 1024, 768], broadcast_dims=[2])
    T5 = fd.ops.view(T2, original_shape=[1024, 768], new_shape=[1, 1024, 768])
    T6 = fd.ops.cast(T5, dtype=DataType.Float)
    S7 = fd.define_constant(0.500000)
    T8 = fd.ops.mul(T6, S7)
    S9 = fd.define_constant(0.707107)
    T10 = fd.ops.mul(T6, S9)
    T11 = fd.ops.erf(T10)
    S12 = fd.define_constant(1.00000)
    T13 = fd.ops.add(T11, S12)
    T14 = fd.ops.mul(T8, T13)
    T15 = fd.ops.cast(T14, dtype=DataType.Half)
    T16 = fd.ops.cast(T15, dtype=DataType.Float)
    T17, T18 = fd.ops.var_mean(T16, axes=[2], correction=0, keepdim=False)
    T19 = fd.ops.broadcast_in_dim(T17, output_shape=[1, 1024, 1], broadcast_dims=[0, 1])
    T20 = fd.ops.broadcast_in_dim(T18, output_shape=[1, 1024, 1], broadcast_dims=[0, 1])
    S21 = fd.define_constant(1.00000e-05)
    T22 = fd.ops.add(T19, S21)
    T23 = fd.ops.broadcast_in_dim(T20, output_shape=[1, 1024, 768], broadcast_dims=[0, 1, 2])
    T24 = fd.ops.rsqrt(T22)
    T25 = fd.ops.sub(T16, T23)
    T26 = fd.ops.broadcast_in_dim(T24, output_shape=[1, 1024, 768], broadcast_dims=[0, 1, 2])
    T27 = fd.ops.mul(T25, T26)
    T28 = fd.ops.mul(T27, T3)
    T29 = fd.ops.add(T28, T4)
    T30 = fd.ops.cast(T29, dtype=DataType.Float)
    T31 = fd.ops.cast(T30, dtype=DataType.Half)
    T32 = fd.ops.view(T31, original_shape=[1, 1024, 768], new_shape=[1024, 768])
    fd.add_output(T5)
    fd.add_output(T16)
    fd.add_output(T20)
    fd.add_output(T24)
    fd.add_output(T32)

fs = Fusion()

with FusionDefinition(fs) as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn(768, device='cuda', dtype=torch.float32),
    torch.randn(768, device='cuda', dtype=torch.float32),
    torch.randn(1024, 768, device='cuda', dtype=torch.float16),
]

for _ in range(5):
    out = fs.execute(inputs)

```

Error:

```
RuntimeError: Rfactor replay recieved an axis outside the number of dims in the tensor, acceptable inclusive range is -4 to 3
```

### Versions

`csarofeen/torchbenchPerf`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes https://github.com/csarofeen/pytorch/issues/2094

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 1]
```diff
@@ -161,6 +161,8 @@ void IrPrinter::handle(const TensorView* tv) {
     case MemoryType::Local:
       os_ << "_l";
       break;
+    default:
+      TORCH_INTERNAL_ASSERT(false, "Unknown tensor memory type.");
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp:165]: Unrelated but I think it is a good idea to have it

**Review by csarofeen — APPROVED** [review]
LGTM

**Comment by naoyam** [comment]
So, is this an alternative fix for the problem you previously proposed the "mirror" approach? How does having dummy outputs solve the problem?

**Comment by zasdfgbnm** [comment]
> So, is this an alternative fix for the problem you previously proposed the "mirror" approach? How does having dummy outputs solve the problem?

Yes, it is. Let's say we had
```
T0[I]
T1[b b I] = broadcast(T0)
T2[b b r] = reduction(T1)
T3[b b b] = broadcast(T2)
T4[b, b, I] = T1 + T3
T5[b, b, r] = reduction(T4)
```

After projection, it becomes
```
T0[I]
T1[b b I] = broadcast(T0)
T2[b b r] = reduction(T1)
T3[b b b] = broadcast(T2)
T6[b b I] = broadcast(T0)
T4[b, b, I] = T6 + T3
T5[b, b, r] = reduction(T4)
```

Then neither the propagation path T2->T3->T4->T5 nor T2->T1->T0->T6->T4->T5 works because they both have missing root domain. But adding `T7 = T1 + T6` creates a new propagation path `T2->T1->T7->T6->T4->T5` which has all root domain information.

(edited)

**Comment by naoyam** [comment]
Oh, I see. Interesting approach. Please have your explanation as a code comment.

I wonder how robust and generic this approach would be, but seems like good enough.

**Review by naoyam — APPROVED** [review]

**Comment by csarofeen** [comment]
If it was directly on iter domains instead of tensor views it would be very robust. The fact that it's on tensor views means that we can only build relationships across tensorviews that can go through a pointwise operation. Since this is replicated tensors exactly, it should be sufficient for all our use cases. Though same day in the future all replays should probably be IterDomain based, not TensorView based.

**Pull request merged**